### PR TITLE
making SummaryDiff serializable

### PR DESCRIPTION
**Pull request by mimoo**
Simply making `SummaryDiff` serializable, so that we can print it in json.

It looks like this:

```json
{
  "target_packages": {
    "changed": [
      {
        "dependency": {
          "name": "dep",
          "version": "0.4.2",
          "crates-io": true
        },
        "changes": {
          "Removed": {
            "old_info": {
              "status": "direct",
              "features": [
                "std"
              ]
            }
          }
        }
      },
      {
        "dependency": {
          "name": "dep",
          "version": "0.4.3",
          "crates-io": true
        },
        "changes": {
          "Added": {
            "info": {
              "status": "direct",
              "features": [
                "std"
              ]
            }
          }
        }
      },
      {
        "dependency": {
          "name": "dep",
          "version": "0.5.0",
          "crates-io": true
        },
        "changes": {
          "Added": {
            "info": {
              "status": "transitive",
              "features": [
                "std"
              ]
            }
          }
        }
      },
      {
        "dependency": {
          "name": "foo",
          "version": "1.2.3",
          "workspace-path": "foo"
        },
        "changes": {
          "Modified": {
            "old_version": null,
            "old_source": null,
            "old_status": null,
            "new_status": "initial",
            "added_features": [
              "feature2"
            ],
            "removed_features": [],
            "unchanged_features": [
              "default",
              "feature1"
            ]
          }
        }
      }
    ],
    "unchanged": {}
  },
  "host_packages": {
    "changed": [
      {
        "dependency": {
          "name": "bar",
          "version": "0.2.0",
          "workspace-path": "dir/bar"
        },
        "changes": {
          "Modified": {
            "old_version": "0.1.0",
            "old_source": null,
            "old_status": "workspace",
            "new_status": "initial",
            "added_features": [],
            "removed_features": [],
            "unchanged_features": [
              "default",
              "feature2"
            ]
          }
        }
      },
      {
        "dependency": {
          "name": "local-dep",
          "version": "1.1.2",
          "path": "../local-dep"
        },
        "changes": {
          "Modified": {
            "old_version": null,
            "old_source": null,
            "old_status": null,
            "new_status": "transitive",
            "added_features": [
              "dep-feature"
            ],
            "removed_features": [],
            "unchanged_features": []
          }
        }
      },
      {
        "dependency": {
          "name": "local-dep",
          "version": "2.0.0",
          "path": "../local-dep-2"
        },
        "changes": {
          "Added": {
            "info": {
              "status": "transitive",
              "features": []
            }
          }
        }
      }
    ],
    "unchanged": {}
  }
}
```

### CODE REVIEW

File: guppy-summaries/src/unit_tests/basic_tests.rs [hunk 1]
```diff
@@ -10,73 +10,8 @@ use semver::Version;
 use std::collections::BTreeSet;
 
 type Summary = SummaryWithMetadata;
-#[test]
```
Comment (mimoo) [inline guppy-summaries/src/unit_tests/basic_tests.rs:13]: the diff looks weird, but I just moved the example inputs outside of the testing functions.

File: guppy-summaries/src/diff.rs [hunk 2]
```diff
@@ -182,8 +183,35 @@ impl<'a> PackageDiff<'a> {
...
+        }
+
+        let changed: Vec<Changed> = self
+            .changed
+            .iter()
+            .map(|(dependency, changes)| Changed {
+                dependency,
+                changes,
+            })
+            .collect();
```
Comment (sunshowers) [inline guppy-summaries/src/diff.rs:204]: Looks good and is the right way to do this, I think.

File: guppy-summaries/src/diff.rs [hunk 3]
```diff
@@ -182,8 +183,35 @@ impl<'a> PackageDiff<'a> {
...
 }
 
+impl<'a> Serialize for PackageDiff<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        #[derive(Serialize)]
+        struct Changed<'a> {
+            dependency: &'a SummaryId,
```
Comment (sunshowers) [inline guppy-summaries/src/diff.rs:193]: Only change is I'd replace "dependency" with "package" here. I'll update your PR with this change. Thanks 💕

**Review by sunshowers — APPROVED** [review]

**Comment by sunshowers** [comment]
Going to release 0.3.1 once CI passes, thanks :)

**Pull request closed without merging**